### PR TITLE
Adds new electronic holdings record

### LINKED ISSUE #102: 856 Electronic Holdings Record Task
Add task to the `bib_records` DAG to generate a FOLIO Holding Records based on values in the MARC records 856 fields. Based on these mapping values contained in https://docs.google.com/spreadsheets/d/1XUSjuysc33NUIJ3SpoYAbTL7_rQojJhGm9KUAlaQD_Q/edit#gid=0, with the following assumptions:

- The last subfield x in the 856 field may contain the Vendor code. This value should be added as a note with a note type of *Vendor* on the holdings record. Each Vendor has a separate FOLIO Electronic Holdings Record
- Remaining subfield x values stored in a Holdings note. 
- Volume information in subfield 3 in the 856 will be stored in the in `electronicAccess.materialsSpecification`

### PR DESCRIPTION
**Pull request by jermnelson**
Fixes #102. 

TODO:
- [x] Add correct `LOCATION` to Electronic holdings tsv
- [x] Troubleshoot mapping errors: 
 ```
[2022-06-16, 20:51:35 UTC] {custom_exceptions.py:52} DATA_ISSUES - RECORD FAILED	ddbc76ae-0977-4f87-93b2-f6cded6db65b	Bib id not in instance id map.	a14240796
[2022-06-16, 20:51:35 UTC] {custom_exceptions.py:51} ERROR - Record failed: Bib id not in instance id map.. See the data issues log for details
```
- [x] Linting errors
- [x] Unit tests

### CODE REVIEW

File: plugins/folio/helpers.py [hunk 1]
```diff
@@ -69,33 +70,99 @@ def move_marc_files_check_tsv(*args, **kwargs) -> str:
...
+            row[field] = None
+        row["CATKEY"] = catkey
+        row["URI"] = "".join(field856.get_subfields("u"))
+        material_type = field856.get_subfields("3")
+        if len(material_type) > 0:
+            row["MAT_SPEC"] = " ".join(material_type)
+        if link_text := field856.get_subfields("y"):
+            row["LINK_TEXT"] = " ".join(link_text)
+        if public_note := field856.get_subfields("z"):
+            row["PUBLIC_NOTE"] = " ".join(public_note)
```
Comment (shelleydoljack) [inline plugins/folio/helpers.py:130]: Are multiple public notes supposed to be mapped to one folio electronicAccess.publicNote field or multiple? If multiple, then probably going to need to join with something other than a space.

Comment (shelleydoljack) [reply plugins/folio/helpers.py:130]: Per Alissa, only one publicNote field is allowed per URL so this seems fine to join with a space.

File: plugins/tests/test_helpers.py [hunk 2]
```diff
@@ -354,3 +357,90 @@ def test_setup_data_logging(mock_logger_file_handler):
...
+        subfields=["z", "table of contents", "u", "http://example.com/"],
+    )
+    assert _add_electronic_holdings(skip_856_field) is False
+
+
+def test_add_electronic_holdings():
+    field_856 = Field(
+        tag="856", indicators=["0", "0"], subfields=["u", "http://example.com/"]
+    )
+    assert _add_electronic_holdings(field_856) is True
```
Comment (shelleydoljack) [inline plugins/tests/test_helpers.py:375]: Oh, ok, so holdings are getting added when indicator2 is 0.

**Review by shelleydoljack — CHANGES_REQUESTED** [review]

File: plugins/folio/helpers.py [hunk 3]
```diff
@@ -69,33 +70,124 @@ def move_marc_files_check_tsv(*args, **kwargs) -> str:
...
+
+
+def _add_electronic_holdings(field856: pymarc.Field) -> bool:
+    if field856.indicator2.startswith("1"):
+        subfield_z = field856.get_subfields("z")
+        subfield_3 = field856.get_subfields("3")
+        subfield_all = " ".join(subfield_z + subfield_3)
+        if full_text_check.match(subfield_all):
+            return False
+    return True
```
[the remark this answers is not in the document]
Comment (jermnelson) [reply plugins/folio/helpers.py:102]: > This method should also return true if 856 indicator2 is 0, unless that is handled elsewhere, although I haven't seen that. The full_text_check against subfields z and 3 is really only for when 856 indicator2 is 1.

Yes, this method will return true if indicator2 is 0.

File: plugins/folio/holdings.py [hunk 4]
```diff
@@ -11,10 +13,66 @@
...
+        migration_task_type="HoldingsCsvTransformer",
+        hrid_handling="preserve001",
+        files=[{"file_name": filename, "suppress": False}],
+        create_source_records=False,
+        call_number_type_map_file_name="call_number_type_mapping.tsv",
+        holdings_map_file_name="holdingsrecord_mapping_electronic.json",
+        location_map_file_name="locations.tsv",
+        holdings_type_uuid_for_boundwiths="",
+        holdings_merge_criteria=["instanceId", "permanentLocationId"],
+        default_call_number_type_name="Library of Congress classification",
```
[the remark this answers is not in the document]
Comment (jermnelson) [reply plugins/folio/holdings.py:48]: Not sure, who would know? @ahafele?

File: plugins/folio/helpers.py [hunk 5]
```diff
@@ -69,33 +70,124 @@ def move_marc_files_check_tsv(*args, **kwargs) -> str:
...
+
+    if len(all001) > 1:
+        for field001 in all001[1:]:
+            field035 = pymarc.Field(
+                tag="035", indicators=["", ""], subfields=["a", field001.data]
+            )
+            record.add_field(field035)
+            record.remove_field(field001)
+
+    return catkey
```
[the remark this answers is not in the document]
Comment (jermnelson) [reply plugins/folio/helpers.py:89]: This has been the going assumption for all of the bib record loads.

File: plugins/folio/holdings.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (shelleydoljack) [reply plugins/folio/holdings.py:48]: I don't actually see that a call number type got assigned so 🤷

**Review by shelleydoljack — APPROVED** [review]

**Pull request merged**